File: Backend/Core/src/main/scala/APIs/CurriculumService.scala
package APIs

import Common.API.API
import Global.CurriculumService
import Objects._

import io.circe.generic.auto._

case class CreateSubjectMessage(
    adminToken: String,
    name: String,
    credits: Int
) extends API[Unit](CurriculumService, "CreateSubjectMessage")

case class CreateRecommendationMessage(
    adminToken: String,
    departmentId: String,
    subjectId: String,
    stage: Stage,
    priority: Int
) extends API[Unit](CurriculumService, "CreateRecommendationMessage")

case class CreateCourseMessage(
    teacherToken: String,
    subjectId: String,
    termId: String,
    location: String,
    schedule: String,
    capacity: Int
) extends API[Unit](CurriculumService, "CreateCourseMessage")

case class CreateCommentMessage(
    studentToken: String,
    courseId: String,
    content: String
) extends API[Unit](CurriculumService, "CreateCommentMessage")

case class QuerySubjectsMessage()
    extends API[List[Subject]](CurriculumService, "QuerySubjectsMessage")

case class QueryCurriculaMessage(departmentId: String)
    extends API[List[Curriculum]](
      CurriculumService,
      "QueryCurriculaMessage"
    )

case class QueryRecommendationsMessage(departmentId: String, stage: Stage)
    extends API[List[Recommendation]](
      CurriculumService,
      "QueryRecommendationsMessage"
    )

case class QueryTeachingsMessage(teacherId: String, termId: String)
    extends API[List[Teaching]](
      CurriculumService,
      "QueryTeachingsMessage"
    )

case class QueryCoursesMessage(subjectId: String, termId: String)
    extends API[List[Course]](CurriculumService, "QueryCoursesMessage")

case class QueryCommentsMessage(courseId: String)
    extends API[List[Comment]](CurriculumService, "QueryCommentsMessage")

case class QueryReviewsMessage(subjectId: String, teacherId: String)
    extends API[List[Review]](CurriculumService, "QueryReviewsMessage")

case class UpdateSubjectMessage(
    adminToken: String,
    subjectId: String,
    name: String,
    credits: Int
) extends API[Unit](CurriculumService, "UpdateSubjectMessage")

case class UpdateCourseMessage(
    teacherToken: String,
    courseId: String,
    location: String,
    schedule: String,
    capacity: Int
) extends API[Unit](CurriculumService, "UpdateCourseMessage")

case class DeleteSubjectMessage(adminToken: String, subjectId: String)
    extends API[Unit](CurriculumService, "DeleteSubjectMessage")

case class DeleteRecommendationMessage(
    adminToken: String,
    recommendationId: String
) extends API[Unit](CurriculumService, "DeleteRecommendationMessage")

case class DeleteCourseMessage(teacherToken: String, courseId: String)
    extends API[Unit](CurriculumService, "DeleteCourseMessage")

case class DeleteCommentMessage(studentToken: String, commentId: String)
    extends API[Unit](CurriculumService, "DeleteCommentMessage")

File: Frontend/services/curriculumService.ts
import { Stage } from "@/types/registerService";
import ApiClient from "./apiClient";
import {
  Course,
  Curriculum,
  Recommendation,
  Review,
  Subject,
  Teaching,
} from "@/types/curriculumService";

const port = 10012;
const client = new ApiClient(port);

const CurriculumService = {
  createSubject: client.api<
    { adminToken: string; name: string; credits: number },
    void
  >("CreateSubjectMessage"),

  createRecommendation: client.api<
    {
      adminToken: string;
      departmentId: string;
      subjectId: string;
      stage: Stage;
      priority: number;
    },
    void
  >("CreateRecommendationMessage"),

  createCourse: client.api<
    {
      teacherToken: string;
      subjectId: string;
      termId: string;
      location: string;
      schedule: string;
      capacity: number;
    },
    void
  >("CreateCourseMessage"),

  createComment: client.api<
    {
      studentToken: string;
      courseId: string;
      content: string;
    },
    void
  >("CreateCommentMessage"),

  querySubjects: client.api<{ child: void }, Subject[]>("QuerySubjectsMessage"),

  queryCurricula: client.api<{ departmentId: string }, Curriculum[]>(
    "QueryCurriculaMessage"
  ),

  queryRecommendations: client.api<
    { departmentId: string; stage: Stage },
    Recommendation[]
  >("QueryRecommendationsMessage"),

  queryTeachings: client.api<{ teacherId: string; termId: string }, Teaching[]>(
    "QueryTeachingsMessage"
  ),

  queryCourses: client.api<{ subjectId: string; termId: string }, Course[]>(
    "QueryCoursesMessage"
  ),

  queryComments: client.api<{ courseId: string }, Comment[]>(
    "QueryCommentsMessage"
  ),

  queryReviews: client.api<{ subjectId: string; teacherId: string }, Review[]>(
    "QueryReviewsMessage"
  ),

  updateSubject: client.api<
    { adminToken: string; subjectId: string; name: string; credits: number },
    void
  >("UpdateSubjectMessage"),

  updateCourse: client.api<
    {
      teacherToken: string;
      courseId: string;
      location: string;
      schedule: string;
      capacity: number;
    },
    void
  >("UpdateCourseMessage"),

  deleteSubject: client.api<{ adminToken: string; subjectId: string }, void>(
    "DeleteSubjectMessage"
  ),

  deleteRecommendation: client.api<
    { adminToken: string; recommendationId: string },
    void
  >("DeleteRecommendationMessage"),

  deleteCourse: client.api<{ teacherToken: string; courseId: string }, void>(
    "DeleteCourseMessage"
  ),

  deleteComment: client.api<{ studentToken: string; commentId: string }, void>(
    "DeleteCommentMessage"
  ),
};

export default CurriculumService;
